Bigmodel

psky1111 opened the pull request:
update huggingface framework based vision transformer for react, LaClip and unicom

psky1111 committed:
fix the resolution problem for clip-vision transformer part and swim transformer

fix the resolution problem for clip-vision transformer part and swim transformer

psky1111 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleClas into develop

psky1111 committed:
Revert "Revert "fix resolution problem for swin transformer and clip vit  (#3021)""

This reverts commit 174db431a82fb168c01b0be03fbb1d822314bbb1.

Update foundation_vit.py

Update foundation_vit.py

Revert "fix resolution problem for swin transformer and clip vit  (#3021)"

This reverts commit 61f748de6767daa2ee8c243146b2204689ecd1ba.

psky1111 commented:
前向对齐验证用代码.包含huggingface framew vision transformer参数转化代码.其中unicom需要额外在patch_embedding类中的conv2d中添加bias项,默认为无bias.
```python

import torch
import paddle
import numpy as np
from reprod_log import ReprodDiffHelper
from reprod_log import ReprodLogger
from paddle import nn
from ppcls.arch.backbone.model_zoo.foundation_vit import React_vit_base_patch32_224,LaCLIP_vit_base_patch32_224,Unicom_vit_base_patch32_224
import open_clip as clip_torch
from ppcls.utils.logger import init_logger
MODEL_torch_param = {
    "react": "pretrain_model_torch/react.pt",
    "laclip": "pretrain_model_torch/laclip.pt",
    "unicom": "pretrain_model_torch/unicom.pt",
}
Model_paddle_param = {
    "react": React_vit_base_patch32_224,
    "laclip":LaCLIP_vit_base_patch32_224,
    "unicom":Unicom_vit_base_patch32_224
}
def torch_to_numpy(name, x: torch.nn.Module):
    try:
        ckpt = torch.load(MODEL_torch_param[name])["state_dict"]
    except:
        ckpt = torch.load(MODEL_torch_param[name])
    x.load_state_dict(ckpt)
    try:
        vision = x.visual
        state_dict_vision = x.visual.state_dict()
    except:
        vision = x
        state_dict_vision = x. state_dict()
    state_numpy = {}
    for k in state_dict_vision.keys():
        state_numpy[k] = state_dict_vision[k].cpu().numpy()
    return vision, state_numpy
def transfer_rule(x:str):
    if x == "conv1.weight":
        return "patch_embed.proj.weight",False
    elif x == "class_embedding":
        return "cls_token",False
    elif x == "positional_embedding":
        return "pos_embed",False
    elif x.find("transformer.resblocks.") != -1:
        x = x.replace("transformer.resblocks.","blocks.")
        if x.find("ln_")!=-1:
            x = x.replace("ln_","norm")
            return x,False
        elif x.find("attn")!=-1:
            if x.find("in_proj_weight")!=-1:
                x = x.replace("in_proj_weight","qkv.weight")
                return x,True
            elif x.find("in_proj_bias")!=-1:
                x = x.replace("in_proj_bias","qkv.bias")
                return x,False
            elif x.find("out_proj.weight")!=-1:
                return x.replace("out_proj.weight","proj.weight"),True
            elif x.find("out_proj.bias")!=-1:
                return x.replace("out_proj.bias","proj.bias"),False
        elif x.find("mlp")!=-1:
            if x.find("c_fc")!=-1:
                if x.find("weight")!=-1:
                    return x.replace("c_fc","fc1"),True
                else:
                    return x.replace("c_fc","fc1"),False
            elif x.find("c_proj")!=-1:
                if x.find("weight")!=-1:
                    return x.replace("c_proj","fc2"),True
                else:
                    return x.replace("c_proj","fc2"),False
    elif x.find("ln_post")!=-1:
        return x.replace("ln_post","norm"),False
    else:
        return x,False
            
            
    
def numpy_to_paddle(state_dict_numpy, x: paddle.nn.Layer):
    state_paddle = {}
    try:
        a = x.visual.state_dict()
    except:
        a = x.state_dict()
    for k in state_dict_numpy.keys():
        numpy_t = state_dict_numpy[k]
        ## linear neeed to transport, already done based on pytorch
        re_key, is_linear = transfer_rule(k)
        if is_linear:
            if re_key.find("weight")!=-1:
                v = paddle.to_tensor(numpy_t,dtype=paddle.float32)
                new_shape = [1, 0] + list(range(2, v.ndim))
                state_paddle[re_key] = v.transpose(new_shape)
        else:
            state_paddle[re_key] = paddle.to_tensor(numpy_t,dtype=paddle.float32)
    try:
        vision = x.visual
    except:
        vision = x
    state_paddle["cls_token"] = state_paddle["cls_token"].reshape((1,1,-1))
    state_paddle["pos_embed"] = state_paddle["pos_embed"].reshape((1,state_paddle["pos_embed"].shape[0],-1))
    vision.set_state_dict(state_paddle)
    return vision
def check(name = "react"):
    torch.manual_seed(0)
    paddle.seed(0)
    init_logger()
    reprod_logger = ReprodLogger()
    vit_b_32_paddle = Model_paddle_param[name]()
    vit_b_32_torch = clip_torch.load_openai_model("ViT-B-32").to("cpu").to(torch.float32) #ViT-B/32
    vision_torch, state_dict_paddle = torch_to_numpy(name, vit_b_32_torch)
    vision_paddle = numpy_to_paddle(state_dict_paddle, vit_b_32_paddle)
    input = torch.ones([1, 3, 224, 224]).numpy()
    vision_torch.eval()
    vision_paddle.eval()
    paddle_out = vision_paddle(paddle.to_tensor(input))
    reprod_logger.add("logits", paddle_out.numpy())
    reprod_logger.save("./result/forward_paddle.npy")
    torch_out = vision_torch(torch.Tensor(input))
    reprod_logger.add("logits", torch_out.detach().numpy())
    reprod_logger.save("./result/forward_torch.npy")
    diff_helper = ReprodDiffHelper()
    torch_info = diff_helper.load_info("./result/forward_torch.npy")
    paddle_info = diff_helper.load_info("./result/forward_paddle.npy")
    # compare result and produce log
    diff_helper.compare_info(torch_info, paddle_info)
    diff_helper.report(
        path="./result/log/forward_diff.log", diff_threshold=0.1)
if __name__ == "__main__":
    check("react")
    check("laclip")
    check("unicom")
```

cuicheng01 commented on ppcls/arch/backbone/model_zoo/foundation_vit.py:
```diff
@@ -834,6 +870,104 @@ def CLIP_vit_base_patch16_224(pretrained=False, use_ssld=False, **kwargs):
     return model
 
 
+def CLIP_vit_base_patch32_224_huggingface(pretrained=False,
```
这里不要出现huggingface，可以写在备注里

psky1111 replied:
好的

psky1111 replied:
我会统一命名为{BIg model}_vit_base_patch32_224

cuicheng01 commented on ppcls/arch/backbone/model_zoo/foundation_vit.py:
```diff
@@ -834,6 +870,104 @@ def CLIP_vit_base_patch16_224(pretrained=False, use_ssld=False, **kwargs):
+    return model
+
+
+def CLIP_vit_base_patch32_224_LaClip(pretrained=False,
```
这个名字应该改为`LaClip_vit_base_patch32_224`

psky1111 replied:
好的

cuicheng01 reviewed (commented):
这里只有LaClip吗？另外两个没有新增吗？

psky1111 committed:
add backbone function

psky1111 committed:
fix static graph problem

psky1111 commented:
预训练参数
链接：https://pan.baidu.com/s/19c2p5uVT3fzyjoZJjk48dA 
提取码：sztx 
--来自百度网盘超级会员V6的分享

psky1111 committed:
remove text encoder framework and add classifier header directly

psky1111 committed:
fix bug in clip when using classifier header

cuicheng01 commented on eval.py:
```diff
@@ -0,0 +1,18 @@
+from __future__ import absolute_import
```
这个py需要去掉

cuicheng01 commented on ppcls/arch/backbone/model_zoo/foundation_vit.py:
```diff
@@ -834,6 +880,141 @@ def CLIP_vit_base_patch16_224(pretrained=False, use_ssld=False, **kwargs):
+
+
+def React_vit_base_patch32_224(pretrained=False, use_ssld=False, **kwargs):
+    model_name = "CLIP_vit_base_patch32_224"
```
这部分代码逻辑需要变一下，所有的模型如CLIP、React、LaCLIP等，名字都可以换成`vit_base_patch32_224`，统一改一下吧

cuicheng01 commented on ppcls/arch/backbone/model_zoo/swin_transformer_v2.py:
```diff
@@ -709,14 +660,30 @@ def __init__(self,
         else:
             self.norm = None
 
+    def maybe_pad(self, pixel_values, height, width):
```
swin_v2为什么需要改动呢？

psky1111 committed:
updated

psky1111 committed:
support embeding

Note, support embed is only for check the model since there is no related text encoder

psky1111 commented:
transfer_tool:
[transfer_script.zip](https://github.com/PaddlePaddle/PaddleClas/files/13839557/transfer_script.zip)

psky1111 committed:
compatible with param transfer

psky1111 committed:
update setting
